gsplat/strategy/default.py

```diff
@@ -0,0 +1,334 @@
+from typing import Any, Callable, DefaultDict, Dict, List, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+
+from gsplat.utils import normalized_quat_to_rotmat
+
+from .base import Strategy
+
+
+class DefaultStrategy(Strategy):
+    """A default strategy that follows the original 3DGS paper:
+
+    "3D Gaussian Splatting for Real-Time Radiance Field Rendering"
+    https://arxiv.org/abs/2308.04079
+
+    If absgrad=True, it will use the absolute gradients for GS splitting, following the
+    AbsGS paper:
+
+    "AbsGS: Recovering Fine Details for 3D Gaussian Splatting"
+    https://arxiv.org/abs/2404.10484
+    """
+
+    def __init__(
+        self,
+        params: torch.nn.ParameterDict,
+        optimizers: List[torch.optim.Optimizer],
+        verbose: bool = False,
+        scene_scale: float = 1.0,
+        # GSs with opacity below this value will be pruned
+        prune_opa: float = 0.005,
+        # GSs with image plane gradient above this value will be split/duplicated
+        grow_grad2d: float = 0.0002,
+        # GSs with scale below this value will be duplicated. Above will be split
+        grow_scale3d: float = 0.01,
+        # GSs with scale above this value will be pruned.
+        prune_scale3d: float = 0.1,
+        # Start refining GSs after this iteration
+        refine_start_iter: int = 500,
+        # Stop refining GSs after this iteration
+        refine_stop_iter: int = 15_000,
+        # Reset opacities every this steps
+        reset_every: int = 3000,
+        # Refine GSs every this steps
+        refine_every: int = 100,
+        # Use absolute gradients for GS splitting
+        absgrad: bool = False,
+        # Whether to use revised opacity heuristic from arXiv:2404.06109
+        # (experimental)
+        revised_opacity: bool = False,
+    ):
+        super().__init__(params, optimizers)
+
+        for key in ["means3d", "scales", "quats", "opacities"]:
+            assert key in self._params, f"{key} is required in params but missing."
+
+        self._verbose = verbose
+        self._scene_scale = scene_scale
+        self._prune_opa = prune_opa
+        self._grow_grad2d = grow_grad2d
+        self._grow_scale3d = grow_scale3d
+        self._prune_scale3d = prune_scale3d
+        self._refine_start_iter = refine_start_iter
+        self._refine_stop_iter = refine_stop_iter
+        self._reset_every = reset_every
+        self._refine_every = refine_every
+        self._absgrad = absgrad
+        self._revised_opacity = revised_opacity
```

Comment:
Oh nice! I'm just not that familiar with dataclass so it didn't ring a bell to me. I'll update this so that the nested config could work. Just curious, are there another advantages that dataclass would bring to us (in this case)?

Reply:
I think the convenient constructor, subclassing convenience, being able to use it as a config object are the main things!

There are other benefits of dataclasses (introspection, hashing, comparison, easy immutability via `frozen=`, etc) but I don't think they really matter here.